src/libcore/macros.rs

```diff
@@ -94,7 +94,28 @@ macro_rules! assert_eq {
                 }
             }
         }
-    })
+    });
+    ($left:expr , $right:expr, $msg:expr) => ({
+        match (&($left), &($right)) {
+            (left_val, right_val) => {
+                if !(*left_val == *right_val) {
+                    panic!("assertion failed: `(left == right)` \
+                           (left: `{:?}`, right: `{:?}`): {}", left_val, right_val, $msg)
```

Comment:
Maybe this was trying to match `unreachable!` which can also be given a single argument which implements `Display` and use that. If we don't want that behaviour and only accept valid format strings I think we can use a single extra rule like:

```rust
($left:expr , $right:expr, $($arg:tt)*) => ({
    match (&($left), &($right)) {
        (left_val, right_val) => {
            if !(*left_val == *right_val) {
                panic!("assertion failed: `(left == right)` \
                       (left: `{:?}`, right: `{:?}`): {}", left_val, right_val, format_args!($($arg)*))
            }
        }
    }
});
```
This will also accept things like `assert_eq!(1, 2, "{0}", 3)` which I think we do want to accept.

(also we definitely don't want to use `concat!` as we'll just run into #30143 again)